neo/Network/P2P/ProtocolHandler.cs

```diff
@@ -21,9 +22,32 @@ public class SetVersion { public VersionPayload Version; }
         public class SetVerack { }
         public class SetFilter { public BloomFilter Filter; }
 
+        private class LimitedSet<T>
+        {
+            private int maxCapacity;
+            private OrderedDictionary orderedDictionary;
+
+            public LimitedSet(int maxCapacity = 100)
+            {
+                if (maxCapacity <= 0) throw new ArgumentOutOfRangeException(nameof(maxCapacity));
+
+                this.maxCapacity = maxCapacity;
+                this.orderedDictionary = new OrderedDictionary(maxCapacity);
+            }
+
+            public bool Add(T item)
+            {
+                if (orderedDictionary.Contains(item)) return false;
+                if (orderedDictionary.Count >= maxCapacity)
+                    orderedDictionary.RemoveAt(0);
```

Comment:
It would be better if it batch removed when above a threshold instead of needing to remove every time once at capacity. Also, it would be best to remove the least recently used item.

Reply:
Batch remove would possible introduce memory leak again for me.

Reply:
Batch remove shouldn’t introduce a memory leak; suppose it batch removed 1/8 of capacity when it reached capacity. Why do you believe it would introduce a memory leak?

Reply:
It would be fine as you said, but it may need a larger size.